File: MyDailyBeat/MyDailyBeat/EVCViewController.m
//
//  EVCViewController.m
//  MyDailyBeat
//
//  Created by Virinchi Balabhadrapatruni on 8/17/14.
//  Copyright (c) 2014 eVerveCorp. All rights reserved.
//

#import "EVCViewController.h"
#import "BankDatabase.h"
#import "HealthDatabase.h"

@interface EVCViewController ()

@end

@implementation EVCViewController

@synthesize api;

@synthesize mTableView;



- (void)viewDidLoad
{
    [super viewDidLoad];
    
    
    options = [NSArray arrayWithObjects:@"Check My Finances", @"Reach Out ...\nI'm Feeling Blue", @"Find a Job", @"Go Shopping", @"Have a Fling", @"Start a Relationship", @"Make Friends", @"Manage My Health", @"Travel", @"Refer a Friend", nil];
    imageNames = [NSArray arrayWithObjects:@"finance", @"phone", @"briefcase", @"cart", @"hearts", @"hearts", @"peeps", @"health", @"plane", @"peeps", nil];
    
    /*UIImage *image2 = [EVCCommonMethods imageWithImage:[UIImage imageNamed:@"search-icon-white.png"] scaledToSize:CGSizeMake(30, 30)];
    CGRect frameimg3 = CGRectMake(0, 0, image2.size.width, image2.size.height);
    UIButton *someButton3 = [[UIButton alloc] initWithFrame:frameimg3];
    [someButton3 setBackgroundImage:image2 forState:UIControlStateNormal];
    [someButton3 addTarget:self action:@selector(searchGroups)
         forControlEvents:UIControlEventTouchUpInside];
    [someButton3 setShowsTouchWhenHighlighted:YES];
    UIBarButtonItem *searchButton =[[UIBarButtonItem alloc] initWithCustomView:someButton3];*/

    
    UIImage* image3 = [EVCCommonMethods imageWithImage:[UIImage imageNamed:@"hamburger-icon-white"] scaledToSize:CGSizeMake(30, 30)];
    CGRect frameimg = CGRectMake(0, 0, image3.size.width, image3.size.height);
    UIButton *someButton = [[UIButton alloc] initWithFrame:frameimg];
    [someButton setBackgroundImage:image3 forState:UIControlStateNormal];
    [someButton addTarget:self action:@selector(showMenu)
         forControlEvents:UIControlEventTouchUpInside];
    [someButton setShowsTouchWhenHighlighted:YES];
    
    UIBarButtonItem *menuButton =[[UIBarButtonItem alloc] initWithCustomView:someButton];
    NSArray *rightItems = [NSArray arrayWithObjects:menuButton, nil];
    self.navigationItem.rightBarButtonItems = rightItems;
    
    UIImage* image4 = [EVCCommonMethods imageWithImage:[UIImage imageNamed:@"profile-icon-white"] scaledToSize:CGSizeMake(30, 30)];
    CGRect frameimg2 = CGRectMake(0, 0, image4.size.width, image4.size.height);
    UIButton *someButton2 = [[UIButton alloc] initWithFrame:frameimg2];
    [someButton2 setBackgroundImage:image4 forState:UIControlStateNormal];
    [someButton2 addTarget:self action:@selector(showProfile)
         forControlEvents:UIControlEventTouchUpInside];
    [someButton2 setShowsTouchWhenHighlighted:YES];
    
    UIBarButtonItem *profileButton =[[UIBarButtonItem alloc] initWithCustomView:someButton2];
    self.navigationItem.leftBarButtonItem = profileButton;
    
    self.mTableView.dataSource = self;
    self.mTableView.delegate = self;
    api = [RestAPI getInstance];
    NSString *name = [api getCurrentUser].name;
    NSArray *fields = [name componentsSeparatedByString:@" "];
    [self navigationItem].title = [NSString stringWithFormat:@"Welcome %@!", fields[0]];
    
    
    [[UITabBarItem appearance] setTitleTextAttributes:@{ UITextAttributeTextColor :[UIColor whiteColor] } forState:UIControlStateSelected];
    [self.navigationController.navigationBar setTitleTextAttributes:@{NSForegroundColorAttributeName : [UIColor whiteColor]}];
    
    BankDatabase *db = [BankDatabase database];
    NSArray *arr = [db bankInfos];
    if ([arr count] == 0) {
        dispatch_queue_t queue = dispatch_queue_create(APP_ID_C_STRING, NULL);
        dispatch_async(queue, ^{
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.view makeToast:@"Initializing database..." duration:3.5 position:@"bottom"];
                [self.view makeToastActivity];
            });
            // initialize db
            NSArray *temp = TOP_TEN_BANKS;
            for (int i = 0; i < [temp count]; ++i) {
                NSString *tempString = [temp objectAtIndex: i];
                if ([[RestAPI getInstance] doesAppExistWithTerm:tempString andCountry:@"US"]) {
                    VerveBankObject *bank = [[RestAPI getInstance] getBankInfoForBankWithName:tempString inCountry:@"US"];
                    BankInfo *info = [[BankInfo alloc] initWithUniqueId:bank.uniqueID name:bank.appName appURL:bank.appStoreListing iconURL:bank.appIconURL];
                    [db insertIntoDatabase:info];
                }
            }
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.view hideToastActivity];
                [self.view makeToast:@"Database Initialization Complete" duration:3.5 position:@"bottom"];
            });
        });
    }
    
    if ([[NSUserDefaults standardUserDefaults] boolForKey:@"FirstTimeLogin"]) {
        EVCFirstTimeSetupViewController *first = [[EVCFirstTimeSetupViewController alloc] initWithNibName:@"EVCFirstTimeSetupViewController" bundle:nil];
        [self.navigationController pushViewController:first animated:YES];
    }
    
    
}

- (void) dismissGroupSearchViewController: (EVCGroupSearchViewViewController *) controller {
    [controller dismissViewControllerAnimated:YES completion:nil];
}

- (void)searchGroups {
    /*EVCGroupSearchViewViewController *searchController = [[EVCGroupSearchViewViewController alloc] init];
    searchController.delegate = self;
    [self.sideMenuViewController setContentViewController:[[UINavigationController alloc] initWithRootViewController:searchController] animated:YES];*/
    
}

- (void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    switch (indexPath.section) {
        case 1:
            switch (indexPath.row) {
                case 4: {
                    EVCFlingViewController *fling = [[EVCFlingViewController alloc] initWithNibName:@"EVCFlingViewController" bundle:nil andInMode:1];
                    [[NSUserDefaults standardUserDefaults] setInteger:1 forKey:@"REL_MODE"];
                    [self.sideMenuViewController setContentViewController:[[UINavigationController alloc] initWithRootViewController:fling] animated:YES];
                }
                    
                    break;
                case 5: {
                    EVCFlingViewController *fling = [[EVCFlingViewController alloc] initWithNibName:@"EVCFlingViewController" bundle:nil andInMode:2];
                    [[NSUserDefaults standardUserDefaults] setInteger:2 forKey:@"REL_MODE"];
                    [self.sideMenuViewController setContentViewController:[[UINavigationController alloc] initWithRootViewController:fling] animated:YES];
                }
                    
                    break;
                case 6: {
                    EVCFlingViewController *fling = [[EVCFlingViewController alloc] initWithNibName:@"EVCFlingViewController" bundle:nil andInMode:0];
                    [[NSUserDefaults standardUserDefaults] setInteger:0 forKey:@"REL_MODE"];
                    [self.sideMenuViewController setContentViewController:[[UINavigationController alloc] initWithRootViewController:fling] animated:YES];
                }
                    
                    break;
                case 3:
                {
                    EVCShoppingViewController *shop = [[EVCShoppingViewController alloc] initWithNibName:@"EVCShoppingViewController" bundle:nil];
                    [self.sideMenuViewController setContentViewController:[[UINavigationController alloc] initWithRootViewController:shop] animated:YES];
                }
                    break;
                case 1:
                {
                     EVCFeelingBlueTabViewController *fb = [[EVCFeelingBlueTabViewController alloc] initWithNibName:@"EVCFeelingBlueTabViewController" bundle:nil];
                    [self.sideMenuViewController setContentViewController:[[UINavigationController alloc] initWithRootViewController:fb] animated:YES];
                }
                    break;
                    
                case 0:
                {
                    EVCFinanceViewController *finance = [[EVCFinanceViewController alloc] initWithNibName:@"EVCFinanceViewController" bundle:nil];
                    [self.sideMenuViewController setContentViewController:[[UINavigationController alloc] initWithRootViewController:finance] animated:YES];
                    break;
                }
                case 8: {
                    EVCTravelTabViewController *travel = [[EVCTravelTabViewController alloc] init];
                    [self.sideMenuViewController setContentViewController:[[UINavigationController alloc] initWithRootViewController:travel] animated:YES];
                    break;
                }
                    
                case 2: {
                    EVCJobsTabViewController *jobs = [[EVCJobsTabViewController alloc] init];
                    [self.sideMenuViewController setContentViewController:[[UINavigationController alloc] initWithRootViewController:jobs] animated:YES];
                    break;
                }
                    
                case 7: {
                    EVCHealthViewController *health = [[EVCHealthViewController alloc] initWithNibName:@"EVCHealthViewController" bundle:nil];
                    [self.sideMenuViewController setContentViewController:[[UINavigationController alloc] initWithRootViewController:health] animated:YES];
                }
                    break;
                case 9: {
                    DLAVAlertView *alert = [[DLAVAlertView alloc] initWithTitle:@"Refer a Friend" message:@"Enter the name and email address of the person you wish to invite to join MyDailyBeat." delegate:nil cancelButtonTitle:@"Cancel" otherButtonTitles:@"Invite", nil];
                    [alert addTextFieldWithText:@"" placeholder:@"Name"];
                    [alert addTextFieldWithText:@"" placeholder:@"E-mail Address"];
                    [alert setKeyboardType:UIKeyboardTypeEmailAddress ofTextFieldAtIndex:1];
                    [alert setAutoCapitalizationType:UITextAutocapitalizationTypeNone ofTextFieldAtIndex:1];
                    [alert setAutoCorrectionType:UITextAutocorrectionTypeNo ofTextFieldAtIndex:0];
                    [alert setAutoCorrectionType:UITextAutocorrectionTypeNo ofTextFieldAtIndex:1];
                    [alert showWithCompletion:^(DLAVAlertView *alertView, NSInteger buttonIndex) {
                        switch (buttonIndex) {
                            case 0:
                                // do nothing
                                break;
                            case 1: {
                                NSString *name = [alertView textFieldTextAtIndex:0];
                                NSString *email = [alertView textFieldTextAtIndex:1];
                                dispatch_queue_t queue = dispatch_queue_create(APP_ID_C_STRING, NULL);
                                dispatch_async(queue, ^{
                                    dispatch_async(dispatch_get_main_queue(), ^{
                                        [self.view makeToastActivity];
                                    });
                                    BOOL result = [[RestAPI getInstance] sendReferralFromUser:[[RestAPI getInstance] getCurrentUser] toPersonWithName:name andEmail:email];
                                    dispatch_async(dispatch_get_main_queue(), ^{
                                        [self.view hideToastActivity];
                                        if (result) {
                                            [self.view makeToast:@"Referral sent successfully!" duration:3.5 position:@"bottom"];
                                        } else {
                                            [self.view makeToast:@"Could not send referral." duration:3.5 position:@"bottom"];
                                        }
                                        
                                    });
                                });
                            }
                                break;
                                
                            default:
                                break;
                        }
                    }];
                }
                    break;
                    
                    
                default:
                    break;
            }
            break;
            
        default:
            break;
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    static NSString *CellIdentifier = @"Celler";
    UITableViewCell *cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
    
    
    
    switch (indexPath.section) {
        case 0: {
            
            UILabel *textLabel = [[UILabel alloc] initWithFrame:cell.contentView.bounds];
            
            
            textLabel.text = @"What would you like to do?";
            textLabel.textAlignment = NSTextAlignmentCenter;
            textLabel.font = [UIFont fontWithName:@"HelveticaNeue-Bold" size:18];
            textLabel.lineBreakMode = UILineBreakModeWordWrap;
            textLabel.numberOfLines = 0;
            [cell addSubview:textLabel];
            cell.selectionStyle = UITableViewCellSelectionStyleNone;
            break;
        }
        case 1: {
            
            
            cell.textLabel.lineBreakMode = UILineBreakModeWordWrap;
            cell.textLabel.numberOfLines = 0;
            cell.textLabel.text = [options objectAtIndex:indexPath.row];
            cell.textLabel.textAlignment = NSTextAlignmentLeft;
            cell.textLabel.font = [UIFont fontWithName:@"HelveticaNeue" size:18];
            UIImage *icon = [UIImage imageNamed:[imageNames objectAtIndex:indexPath.row]];
            cell.imageView.image = [EVCCommonMethods imageWithImage:icon scaledToSize:CGSizeMake(30, 30)];
            break;
        }
            
    }
    
    //cell.contentView.backgroundColor = UIColorFromHex(0xEEE2BE);
    
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    switch (indexPath.section) {
        case 0:
            return 35;
            break;
            
        case 1:
            if ([(NSString *)[options objectAtIndex:indexPath.row] isEqualToString:@"Reach Out ...\nI'm Feeling Blue"]) {
                return 43;
            } else {
                return 42;
            }
            break;
    }
    return 35;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 2;
}

- (NSInteger)tableView:(UITableView *)tableView
 numberOfRowsInSection:(NSInteger)section {
    switch (section) {
        case 0:
            return 1;
        case 1:
            return [options count];
        default:
            return 1;
    }
}

- (void) showMenu {
    [self.sideMenuViewController presentRightMenuViewController];
}

- (void) showProfile {
    [self.sideMenuViewController presentLeftMenuViewController];
}

@end

File: MDB_Swift/MyDailyBeat/MyDailyBeat/src/EVCViewController.swift

//
//  EVCViewController.swift
//  MyDailyBeat
//
//  Created by Virinchi Balabhadrapatruni on 8/17/14.
//  Copyright (c) 2014 eVerveCorp. All rights reserved.
//
import UIKit
import Toast_Swift
import API
import DLAlertView
import RESideMenu
import GBVersionTracking
class EVCViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {
    var options = [String]()
    var imageNames = [String]()

    @IBOutlet var mTableView: UITableView!
    var api: RestAPI!


    override func viewDidLoad() {
        super.viewDidLoad()
        options = ["Check My Finances", "Reach Out ...\nI'm Feeling Blue", "Find a Job", "Go Shopping", "Have a Fling", "Start a Relationship", "Make Friends", "Manage My Health", "Travel", "Refer a Friend", "Volunteering"]
        imageNames = ["finance", "phone", "briefcase", "cart", "fling", "hearts", "peeps", "health", "plane", "refer", "hands"]
        self.mTableView.dataSource = self
        self.mTableView.delegate = self
        self.mTableView.separatorStyle = .none
        api = RestAPI.getInstance()
        let name: String = api.getCurrentUser().name
        var fields = name.components(separatedBy: " ")
        self.navigationItem.title = "Welcome \(fields[0])!"
        UITabBarItem.appearance().setTitleTextAttributes([NSForegroundColorAttributeName: UIColor.white], for: .selected)
        self.navigationController?.navigationBar.titleTextAttributes = [NSForegroundColorAttributeName: UIColor.white]
        let arr = DataManager.getBanks()
        let arr2 = DataManager.getHealthPortals()
        let arr3 = DataManager.getPrescriptionProviders()
        if arr.isEmpty || arr2.isEmpty || arr3.isEmpty {
            DispatchQueue.global().async(execute: {() -> Void in
                DispatchQueue.main.async(execute: {() -> Void in
                    self.view.makeToast("Fine Tuning Your Experience...", duration: 3.5, position: .bottom)
                    self.view.makeToastActivity(ToastPosition.center)
                })
                    // initialize db
                if arr.isEmpty {
                    var temp = TOP_TEN_BANKS
                    for i in 0..<temp.count {
                        let tempString: String = temp[i]
                        if RestAPI.getInstance().doesAppExist(withTerm: tempString, andCountry: "US") {
                            let bank: VerveBankObject = RestAPI.getInstance().getBankInfoForBank(withName: tempString, inCountry: "US")
                            let info = BankInfo(uniqueId: bank.uniqueID, name: bank.appName, appURL: bank.appStoreListing, iconURL: bank.appIconURL)
                            DataManager.insertBank(info)
                        }
                    }
                }
                
                if arr2.isEmpty {
                    var temp = HEALTH_PORTALS
                    for i in 0..<temp.count {
                        let tempString: String = temp[i]
                        let logoURL = HEALTH_PORTAL_LOGO_URLS[i]
                        let info = HealthInfo(uniqueId: i, url: tempString, logoURL: logoURL)
                        DataManager.insertHealthPortal(info)
                    }
                }
                
                if arr3.isEmpty {
                    var temp = PRESCRIP_PROVIDERS
                    for i in 0..<temp.count {
                        let tempString: String = temp[i]
                        let logoURL = PRESCRIP_PROVIDER_LOGO_URLS[i]
                        let info = PrescripProviderInfo(uniqueId: i, url: tempString, logoURL: logoURL)
                        DataManager.insertPrescriptionProvider(info)
                    }
                }
                
                DispatchQueue.main.async(execute: {() -> Void in
                    self.view.hideToastActivity()
                    UserDefaults.standard.set(false, forKey: "LOAD_BANK_IMAGES")
                    self.view.makeToast("Fine Tuning Complete", duration: 3.5, position: .bottom)
                    if GBVersionTracking.isFirstLaunchEver() {
                        self.showModalSegue(withIdentifier: "FirstTimeSetupSegue", andSender: self)
                    }
                })
            })
        }
        
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.navigationController?.setNavigationBarHidden(true, animated: false)
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let dest = segue.destination as? EVCFlingViewController {
            let mode = sender as! Int
            dest.mode = REL_MODE(rawValue: mode)!
        }
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        switch indexPath.section {
            case 1:
                switch indexPath.row {
                    case 4:
                        self.performSegue(withIdentifier: "RelationshipSegue", sender: 1)
                    case 5:
                        self.performSegue(withIdentifier: "RelationshipSegue", sender: 2)
                    case 6:
                        self.performSegue(withIdentifier: "RelationshipSegue", sender: 0)
                    case 3:
                        self.performSegue(withIdentifier: "ShoppingSegue", sender: nil)
                    case 1:
                        self.performSegue(withIdentifier: "FeelingBlueSegue", sender: nil)
                    case 0:
                        self.performSegue(withIdentifier: "FinanceSegue", sender: nil)
                    case 8:
                        self.performSegue(withIdentifier: "TravelSegue", sender: nil)
                    case 2:
                        self.performSegue(withIdentifier: "JobsSegue", sender: nil)
                    case 7:
                        self.performSegue(withIdentifier: "HealthSegue", sender: nil)
                    case 10:
                        self.performSegue(withIdentifier: "VolunteeringSegue", sender: nil)
                    case 9:
                        let alert = UIAlertController(title: "Refer a Friend", message: "Enter the name and email address of the person you wish to invite to join MyDailyBeat.", preferredStyle: .alert)
                        alert.addTextField(configurationHandler: { (nameField) in
                            nameField.placeholder = "Name"
                            nameField.autocapitalizationType = .none
                            nameField.autocorrectionType = .no
                        })
                        alert.addTextField(configurationHandler: { (emailField) in
                            emailField.placeholder = "E-mail Address"
                            emailField.autocapitalizationType = .none
                            emailField.autocorrectionType = .no
                            emailField.keyboardType = .emailAddress
                        })
                        let action = UIAlertAction(title: "Cancel", style: .cancel, handler: { (action) in
                            // do nothing here
                        })
                        let ok = UIAlertAction(title: "Invite", style: .default, handler: { (inviteAction) in
                            if let fields = alert.textFields {
                                let name = fields[0].text
                                let email = fields[1].text
                                DispatchQueue.global().async(execute: {() -> Void in
                                    DispatchQueue.main.async(execute: {() -> Void in
                                        self.view.makeToastActivity(ToastPosition.center)
                                    })
                                    let result: Bool = RestAPI.getInstance().sendReferral(from: RestAPI.getInstance().getCurrentUser(), toPersonWithName: name!, andEmail: email!)
                                    DispatchQueue.main.async(execute: {() -> Void in
                                        self.view.hideToastActivity()
                                        if result {
                                            self.view.makeToast("Referral sent successfully!", duration: 3.5, position: .bottom)
                                        }
                                        else {
                                            self.view.makeToast("Could not send referral.", duration: 3.5, position: .bottom)
                                        }
                                    })
                                })
                            }
                        })
                        alert.addAction(action)
                        alert.addAction(ok)
                        self.present(alert, animated: true, completion: nil)
                    default:
                        break
                }

            default:
                break
        }

    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let CellIdentifier: String = "Celler"
        let cell = UITableViewCell(style: .default, reuseIdentifier: CellIdentifier)
        switch indexPath.section {
            case 0:
                let textLabel = UILabel(frame: cell.contentView.bounds)
                textLabel.text = "What would you like to do?"
                textLabel.textAlignment = .center
                textLabel.font = UIFont(name: "HelveticaNeue-Bold", size: CGFloat(18))
                textLabel.lineBreakMode = .byWordWrapping
                textLabel.numberOfLines = 0
                cell.addSubview(textLabel)
                cell.selectionStyle = .none
            case 1:
                cell.textLabel?.lineBreakMode = .byWordWrapping
                cell.textLabel?.numberOfLines = 0
                cell.textLabel?.text = options[indexPath.row]
                cell.textLabel?.textAlignment = .left
                cell.textLabel?.font = UIFont(name: "HelveticaNeue", size: CGFloat(18))
                let icon = UIImage(named: imageNames[indexPath.row])
                cell.imageView?.image = EVCCommonMethods.image(with: icon!, scaledTo: CGSize(width: CGFloat(30), height: CGFloat(30)))
        default:
            break
        }

        //cell.contentView.backgroundColor = UIColor(netHex: 0xEEE2BE);
        return cell
    }

    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        switch indexPath.section {
            case 0:
                return 35
            case 1:
                if options[indexPath.row] == "Reach Out ...\nI'm Feeling Blue" {
                    return 43
                } else {
                    return 42
                }
        default:
            break
        }

        return 35
    }

    func numberOfSections(in tableView: UITableView) -> Int {
        return 2
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        switch section {
            case 0:
                return 1
            case 1:
                return options.count
            default:
                return 1
        }

    }
}
